packages/core/lib/dlc/finance/CsoInfo.ts

```diff
@@ -0,0 +1,106 @@
+import {
+  ContractDescriptorV1,
+  ContractInfo,
+  ContractInfoV0,
+  DigitDecompositionEventDescriptorV0,
+  MessageType,
+  PayoutFunctionV0,
+  PolynomialPayoutCurvePiece,
+} from '@node-dlc/messaging';
+import { Value } from '@node-lightning/bitcoin';
+
+import { UNIT_MULTIPLIER } from './Builder';
+import { HasContractInfo, HasType } from './OptionInfo';
+
+export interface CsoInfo {
+  maxLoss: Value;
+  maxGain: Value;
+  minPayout: bigint;
+  maxPayout: bigint;
+  contractSize: Value;
+  offerCollateral: Value;
+}
+
+export const getCsoInfoFromContractInfo = (
+  _contractInfo: ContractInfo,
+): CsoInfo => {
+  if (_contractInfo.type !== MessageType.ContractInfoV0)
+    throw Error('Only ContractInfoV0 currently supported');
+
+  const contractInfo = _contractInfo as ContractInfoV0;
+  if (contractInfo.contractDescriptor.type !== MessageType.ContractDescriptorV1)
+    throw Error('Only ContractDescriptorV1 currently supported');
+
+  const oracleInfo = contractInfo.oracleInfo;
+
+  const eventDescriptor = oracleInfo.announcement.oracleEvent
+    .eventDescriptor as DigitDecompositionEventDescriptorV0;
+
+  if (
+    oracleInfo.announcement.oracleEvent.eventDescriptor.type !==
+    MessageType.DigitDecompositionEventDescriptorV0
+  )
+    throw Error('Only DigitDecompositionEventDescriptorV0 currently supported');
+
+  const contractDescriptor = contractInfo.contractDescriptor as ContractDescriptorV1;
+  if (contractDescriptor.payoutFunction.type !== MessageType.PayoutFunctionV0)
+    throw Error('Only PayoutFunctionV0 currently supported');
+
+  const payoutFunction = contractDescriptor.payoutFunction as PayoutFunctionV0;
+  if (payoutFunction.pieces.length === 0)
+    throw Error('PayoutFunction must have at least one PayoutCurvePiece');
+  if (payoutFunction.pieces.length < 3)
+    throw Error('Must be at least 3 PayoutCurvePieces');
+
+  const payoutCurvePiece = payoutFunction.pieces[0]
+    .payoutCurvePiece as PolynomialPayoutCurvePiece;
+  if (payoutCurvePiece.type !== MessageType.PolynomialPayoutCurvePiece)
+    throw Error('Must be PolynomialPayoutCurvePiece');
+
+  const initialPiece = payoutFunction.pieces[0];
+  const finalPiece = payoutFunction.pieces[1];
+
+  const minPayout = initialPiece.endpointPayout;
+  const maxPayout = finalPiece.endpointPayout;
+
+  const startOutcome = initialPiece.endpoint;
+  const endOutcome = finalPiece.endpoint;
```

Comment:
I think we should add a validation somewhere here to ensure it's actually a CSO contractInfo.

To do that we should check that 
1. There are exactly 3 **pieces** in the payoutFunction.
2. `payoutFunction.endpoint0 === payoutFunction.pieces[0].endpoint` (i.e. the first piece is a straight line)
3. `payoutFunction.pieces[1].endpoint === payoutFunction.pieces[2].endpoint` (i.e. the last piece is a straight line)

![image](https://user-images.githubusercontent.com/16860528/173416830-23622995-2fdd-4430-b772-1fddc2384ade.png)


<img width="484" alt="image" src="https://user-images.githubusercontent.com/16860528/173416565-8ec05522-5f05-4ebf-897a-801eb0406569.png">

What you have is technically correct, but if a different CSO curve gets passed in without the above validations ^ then things would break.

Would also love to hear @zpv's thoughts on this.

Reply:
Great points @dillionverma !

Yep, we should be doing validation on the exact shape of the payout curve. (Flat line at maxLoss, sloped linear line from maxLoss to maxGain, flat line at maxGain)

Reply:
Thanks @dillionverma and @zpv on the detailed feedback here. 

I added a new function `validateCsoPayoutFunction` which has the following checks

- Payout Function piece length is 3 https://github.com/AtomicFinance/node-dlc/pull/177/commits/31b43c0df9c0c52cfcf30b111303eff59365c3a7#diff-3d2004e549ee1179986e0b232a714f25ab2d4d44aec0464603d321b0b8469105R127-R130
- PolynomialPayoutCurvePiece type check https://github.com/AtomicFinance/node-dlc/pull/177/commits/31b43c0df9c0c52cfcf30b111303eff59365c3a7#diff-3d2004e549ee1179986e0b232a714f25ab2d4d44aec0464603d321b0b8469105R132-R135
- Payout validation https://github.com/AtomicFinance/node-dlc/pull/177/commits/31b43c0df9c0c52cfcf30b111303eff59365c3a7#diff-3d2004e549ee1179986e0b232a714f25ab2d4d44aec0464603d321b0b8469105R158-R180
  - Flat line at `maxLoss`
  - Sloped linear line from `maxLoss` to `maxGain`
  - Flat line at `maxGain`
- Event Outcome validation https://github.com/AtomicFinance/node-dlc/pull/177/commits/31b43c0df9c0c52cfcf30b111303eff59365c3a7#diff-3d2004e549ee1179986e0b232a714f25ab2d4d44aec0464603d321b0b8469105R140-R156

The rational behind this logic is based on looking at a CSO Payout Function

```
payoutFunction PayoutFunctionV0 {
  type: 42790,
  endpoint0: 0n,
  endpointPayout0: 60000000n,
  extraPrecision0: 0,
  pieces: [
    {
      payoutCurvePiece: [PolynomialPayoutCurvePiece],
      endpoint: 800000n,
      endpointPayout: 60000000n,
      extraPrecision: 0
    },
    {
      payoutCurvePiece: [PolynomialPayoutCurvePiece],
      endpoint: 1200000n,
      endpointPayout: 100000000n,
      extraPrecision: 0
    },
    {
      payoutCurvePiece: [PolynomialPayoutCurvePiece],
      endpoint: 2097151n,
      endpointPayout: 100000000n,
      extraPrecision: 0
    }
  ],
}
```

`endpoint` should always be increasing
`payout` should follow the flat, ascending, flat pattern

inside of `PolynomialPayoutCurvePiece` each one has 2 points

Initial
```
payoutCurvePiece PolynomialPayoutCurvePiece {
  type: 42792,
  points: [
    { eventOutcome: 0n, outcomePayout: 60000000n, extraPrecision: 0 },
    {
      eventOutcome: 800000n,
      outcomePayout: 60000000n,
      extraPrecision: 0
    }
  ]
}
```

Middle
```
payoutCurvePiece PolynomialPayoutCurvePiece {
  type: 42792,
  points: [
    {
      eventOutcome: 800000n,
      outcomePayout: 60000000n,
      extraPrecision: 0
    },
    {
      eventOutcome: 1200000n,
      outcomePayout: 100000000n,
      extraPrecision: 0
    }
  ]
}
```

Final
```
payoutCurvePiece PolynomialPayoutCurvePiece {
  type: 42792,
  points: [
    {
      eventOutcome: 1200000n,
      outcomePayout: 100000000n,
      extraPrecision: 0
    },
    {
      eventOutcome: 2097151n,
      outcomePayout: 100000000n,
      extraPrecision: 0
    }
  ]
}
```

And we can see, `PolynomialPayoutCurvePiece` for CSO should follow similar principles to base `payoutFunction` validation

`eventOutcome` should always be increasing
`outcomePayout` should follow the flat, ascending, flat pattern